Library/Homebrew/utils/formatter.rb

```diff
@@ -32,7 +32,12 @@ def error(string, label: nil)
   end
 
   def wrap(s, width = 172)
-    s.gsub(/(.{1,#{width}})(\s+|\Z)/, "\\1\n")
+    col = 43
+    indent = width - col
+    s.gsub(/(?<=\S) *\n(?=\S)/, " ")
+     .gsub(/^( +-.+  +(?=\S.{#{col}}))(.{1,#{col}})( +|$)\n?/, "\\1\\2\n" + " " * indent)
+     .gsub(/^( {#{indent}}(?=\S.{#{col}}))(.{1,#{col}})( +|$)\n?/, "\\1\\2\n" + " " * indent)
+     .gsub(/(.{1,#{width}})( +|$)\n?/, "\\1\n")
```

Comment:
Could you add a bunch more comments here? At a glance it's very hard to figure out what's going on here. Also: does this wrap at whitespace or in the middle of words?

Reply:
Yes, regular expressions can be like that. 🙃

The first line turns any hard-wrapped paragraphs into a single line. The other three are different takes on the same concept: find a run of characters up to length `n`, then backtrack  (unless the line is already shorter than `n`) until a space is found and replace it with a line break. The algorithm is lifted from this [blog post](https://macromates.com/blog/2006/wrapping-text-with-regular-expressions/), except it doesn't attempt to also line-break words longer than `n`, e.g. URLs.

Reply:
Great, thanks. Would be good to document your comments here inline in Ruby and link the blog post too?